mlflow/gateway/providers/openai.py

```diff
@@ -121,18 +133,214 @@ async def chat_stream(
+                # TODO to support n > 1.
+                assistant_msg = resp["choices"][0]["message"]
+                tool_calls = assistant_msg.get("tool_calls")
+                if tool_calls is None:
+                    if uc_func_calls:
+                        resp["choices"][0]["message"]["content"] = prepend_uc_functions(
+                            resp["choices"][0]["message"]["content"], uc_func_calls
+                        )
+
+                    if user_tool_calls:
+                        resp["choices"][0]["message"]["tool_calls"] = user_tool_calls
+
+                    break
+
+                tool_messages = []
+                for tool_call in tool_calls:  # TODO: should run in parallel
+                    func = tool_call["function"]
+                    parameters = json.loads(func["arguments"])
+                    if function := uc_func_mapping.get(func["name"]):
+                        result = execute_function(
+                            ws=workspace_client,
+                            warehouse_id=warehouse_id,
+                            function=function,
+                            parameters=parameters,
+                        )
+                        if result.value is not None:
+                            tool_message = {
+                                "role": "tool",
+                                "tool_call_id": tool_call["id"],
+                                "content": str(result.value),
+                            }
+                        tool_messages.append(tool_message)
+
+                    if func["name"] in uc_func_mapping:
+                        uc_func_calls.append(
+                            (
+                                {
+                                    "id": tool_call["id"],
+                                    "type": "function",
+                                    "function": {
+                                        "name": func["name"],
+                                        "arguments": func["arguments"],
+                                    },
+                                },
+                                {
+                                    "role": "tool",
+                                    "tool_call_id": tool_call["id"],
+                                    "content": str(result),
+                                },
+                            )
+                        )
+                    else:
+                        should_break = True
+                        user_tool_calls.append(
+                            {
+                                "id": tool_call["id"],
+                                "type": "function",
+                                "function": {
+                                    "name": func["name"],
+                                    "arguments": func["arguments"],
+                                },
+                            }
+                        )
```

Comment:
Can we add an inline comment? It's not clear to me why we decide to break as soon as a non-UC tool is encountered. Shouldn't we be able to switch back and forth between UC tools and non-UC tools?

Reply:
`tool_calls` must be followed by tool messages. If `tool_calls` looks like:

```python

tool_calls = [
    {
        "id": "<uc_func_id>",
        "function": {
            "arguments": "...",
            "name": "uc_func",
        },
        "type": "function",
    },
    {
        "id": "<user_func_id>",
        "function": {
            "arguments": "...",
            "name": "user_func",
        },
        "type": "function",
    },
]

# this must be followed by messages that look like:
[
    {
        "tool_call_id": "<uc_func_id>",
        "role": "tool",
        "name": "uc_func",
        "content": uc_function_response,  # we can get this
    },
    {
        "tool_call_id": "<uc_func_id>",
        "role": "tool",
        "name": "user_func",
        "content": user_func_response,  # but we can't get this
    },
]

```

we can get the result of `uc_function_response` in the server, but we can't get `user_func_response`. Only the client can execute it.

Reply:
@mengxr how is it supposed to work with a mix of uc and user-defined functions?